src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -1699,14 +1706,19 @@ class InstanceCache : public IECore::RefCounted
 			IECore::MurmurHash h = object->hash();
 			arnoldAttributes->hashGeometry( object, h );
 
-			Cache::accessor a;
-			m_cache.insert( a, h );
-			if( !a->second )
+			Cache::const_accessor readAccessor;
+			if( !m_cache.find( readAccessor, h ) )
 			{
-				a->second = convert( object, arnoldAttributes, "instance:" + h.toString() );
+				Cache::accessor writeAccessor;
+				if( m_cache.insert( writeAccessor, h ) )
+				{
+					writeAccessor->second = convert( object, arnoldAttributes, "instance:" + h.toString() );
+				}
+				writeAccessor.release();
+				m_cache.find( readAccessor, h );
```

Comment:
Only Q would be the relative merit of the cost of this vs d.r.y. Could we avoid repetition with a pointer on the outside that we use in the `Instance` constructor?

Reply:
Good point. I've pushed a version which does just that, and realised that it also allows us to drop the lock earlier, so we're not holding the lock when constructing the `Instance`.